--- a/src/backend/tty.rs
+++ b/src/backend/tty.rs
@@ -72,8 +72,12 @@
 
 // When copying from rendering Nvidia dGPU to target iGPU,
 // it only understands X/Abgr and not X/Argb.
-const SUPPORTED_COLOR_FORMATS_10BIT: [Fourcc; 3] =
-    [Fourcc::Abgr2101010, Fourcc::Argb8888, Fourcc::Abgr8888];
+const SUPPORTED_COLOR_FORMATS_10BIT: [Fourcc; 4] = [
+    Fourcc::Abgr2101010,
+    Fourcc::Argb2101010,
+    Fourcc::Argb8888,
+    Fourcc::Abgr8888,
+];
 
 // Smithay should fall back to Xrgb/Xbgr automatically if needed.
 const SUPPORTED_COLOR_FORMATS: [Fourcc; 2] = [Fourcc::Argb8888, Fourcc::Abgr8888];
